code/datums/gas_mixture.dm

```diff
@@ -562,58 +431,151 @@ What are the archived variables for?
+	if(((abs(delta_oxygen) > MINIMUM_AIR_TO_SUSPEND) && (abs(delta_oxygen) >= oxygen * MINIMUM_AIR_RATIO_TO_SUSPEND)) \
+		|| ((abs(delta_carbon_dioxide) > MINIMUM_AIR_TO_SUSPEND) && (abs(delta_carbon_dioxide) >= carbon_dioxide * MINIMUM_AIR_RATIO_TO_SUSPEND)) \
+		|| ((abs(delta_nitrogen) > MINIMUM_AIR_TO_SUSPEND) && (abs(delta_nitrogen) >= nitrogen * MINIMUM_AIR_RATIO_TO_SUSPEND)) \
+		|| ((abs(delta_toxins) > MINIMUM_AIR_TO_SUSPEND) && (abs(delta_toxins) >= toxins * MINIMUM_AIR_RATIO_TO_SUSPEND)) \
+		|| ((abs(delta_sleeping_agent) > MINIMUM_AIR_TO_SUSPEND) && (abs(delta_sleeping_agent) >= sleeping_agent * MINIMUM_AIR_RATIO_TO_SUSPEND)) \
+		|| ((abs(delta_agent_b) > MINIMUM_AIR_TO_SUSPEND) && (abs(delta_agent_b) >= agent_b * MINIMUM_AIR_RATIO_TO_SUSPEND)))
+		return FALSE
+	if(abs(delta_temperature) > MINIMUM_TEMPERATURE_DELTA_TO_SUSPEND)
+		return FALSE
 
-//eg:
-//Tox_PP = get_partial_pressure(gas_mixture.toxins)
-//O2_PP = get_partial_pressure(gas_mixture.oxygen)
+	return TRUE
 
-//Does handle trace gases!
+	///Performs various reactions such as combustion or fusion (LOL)
+	///Returns: TRUE if any reaction took place; FALSE otherwise
+/datum/gas_mixture/proc/react(atom/dump_location)
+	var/reacting = FALSE //set to TRUE if a notable reaction occured (used by pipe_network)
 
-/datum/gas_mixture/proc/get_breath_partial_pressure(gas_pressure)
-	return (gas_pressure * R_IDEAL_GAS_EQUATION * temperature) / BREATH_VOLUME
+	if((agent_b > MINIMUM_MOLE_COUNT) && temperature > 900)
+		if(toxins > MINIMUM_HEAT_CAPACITY && carbon_dioxide > MINIMUM_HEAT_CAPACITY)
+			var/reaction_rate = min(carbon_dioxide * 0.75, toxins * 0.25, agent_b * 0.05)
+
+			carbon_dioxide -= reaction_rate
+			oxygen += reaction_rate
 
+			agent_b -= reaction_rate * 0.05
 
-//Reverse of the above
-/datum/gas_mixture/proc/get_true_breath_pressure(breath_pp)
-	return (breath_pp * BREATH_VOLUME) / (R_IDEAL_GAS_EQUATION * temperature)
+			temperature += (reaction_rate * 20000) / heat_capacity()
+
+			reacting = TRUE
 
-//Mathematical proofs:
-/*
+	if((sleeping_agent > MINIMUM_MOLE_COUNT) && temperature > N2O_DECOMPOSITION_MIN_ENERGY)
+		var/energy_released = 0
+		var/old_heat_capacity = heat_capacity()
+		var/burned_fuel = 0
 
+		burned_fuel = max(0, 0.00002 * (temperature - (0.00001 * (temperature ** 2)))) * sleeping_agent
+		if(sleeping_agent - burned_fuel > 0)
+			sleeping_agent -= burned_fuel
+
+			if(burned_fuel)
+				energy_released += (N2O_DECOMPOSITION_ENERGY_RELEASED * burned_fuel)
+
+				oxygen += burned_fuel * 0.5
+				nitrogen += burned_fuel
+
+				var/new_heat_capacity = heat_capacity()
+				if(new_heat_capacity > MINIMUM_HEAT_CAPACITY)
+					temperature = (temperature * old_heat_capacity + energy_released) / new_heat_capacity
+				reacting = TRUE
+
+	fuel_burnt = 0
+	//Handle plasma burning
+	if((toxins > MINIMUM_MOLE_COUNT) && (oxygen > MINIMUM_MOLE_COUNT) && temperature > FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
```

Comment:
the logic of this is a bit different, and inspired from TG; checking that there's enough plasma and oxygen in the air means that we're skipping on on a bit of math and logic if we wind up with situations where there's not much plasma, but it is very hot.